xcpmd/src/acpi-events.c

```diff
@@ -30,133 +30,406 @@
+    else if (file == NULL)
+        return AC_UNKNOWN;
+
+    fgets(data, sizeof(data), file);
+    fclose(file);
+
+    if (strstr(data, "1"))
+        return ON_AC;
+    else
+        return ON_BATT;
 }
 
-static void handle_battery_info_change_event(void)
-{
-    xcpmd_log(LOG_INFO, "Battery info change event\n");
 
-    if (write_battery_info(NULL) > 0)
-        xenstore_write("1", XS_BATTERY_PRESENT);
+int get_lid_status(void) {
+
+    char data[128];
+    FILE * file;
+    file = fopen(LID_STATE_FILE_PATH, "r");
+
+    //If the file doesn't exist, we're not a laptop.
+    if (file == NULL && errno == ENOENT)
+        return NO_LID;
+    else if (file == NULL)
+        return LID_UNKNOWN;
+
+    fgets(data, sizeof(data), file);
+    fclose(file);
+
+    if (strstr(data, "open"))
+        return LID_OPEN;
+    else if (strstr(data, "closed"))
+        return LID_CLOSED;
     else
-        xenstore_write("0", XS_BATTERY_PRESENT);
+        return LID_UNKNOWN;
+}
+
+
+int get_tablet_status(void) {
+
+    //For now, this is a stub.
+    return NORMAL_MODE;
+
+}
+
+
+static void handle_battery_info_event(int battery_index) {
+
+    char path[256];
+
+    //xcpmd_log(LOG_DEBUG, "Info change event on battery %d\n", battery_index);
+
+    update_battery_info(battery_index);
+    write_battery_info_to_xenstore(battery_index);
+
+    //sprintf(path, "%s%i/%s", XS_BATTERY_PATH, battery_index, XS_BATTERY_PRESENT_LEAF);
+    //xenstore_write((last_info[battery_index].present == YES) ? "1" : "0", path);
+
+    sprintf(path, "%s%i/%s", XS_BATTERY_EVENT_PATH, battery_index, XS_BATTERY_INFO_EVENT_LEAF);
+    xenstore_write("1", path);
 
     notify_com_citrix_xenclient_xcpmd_battery_info_changed(xcdbus_conn, XCPMD_SERVICE, XCPMD_PATH);
 }
 
-static void handle_pbtn_pressed_event(void)
-{
+
+static void handle_battery_status_event(int battery_index) {
+
+    char path[256];
+
+    //xcpmd_log(LOG_DEBUG, "Status change event on battery %d\n", battery_index);
+
+    update_battery_status(battery_index);
+    write_battery_status_to_xenstore(battery_index);
+
+    sprintf(path, "%s%i/%s", XS_BATTERY_EVENT_PATH, battery_index, XS_BATTERY_STATUS_EVENT_LEAF);
+    xenstore_write("1", path);
+
+    //Here for compatibility--should eventually be removed
+    xenstore_write("1", XS_BATTERY_STATUS_CHANGE_EVENT_PATH);
+
+    notify_com_citrix_xenclient_xcpmd_battery_status_changed(xcdbus_conn, XCPMD_SERVICE, XCPMD_PATH);
+}
+
+
+static void handle_lid_event(int status) {
+
+    int lid_status;
+
+    xcpmd_log(LOG_INFO, "Lid change event\n");
+
+    //We may have to check the sysfs if the ACPI string didn't tell us the status.
+    if (status == LID_UNKNOWN)
+        lid_status = get_lid_status();
+    else
+        lid_status = status;
+
+    xenstore_write(lid_status == LID_OPEN ? "open" : "closed", XS_LID_STATE_PATH);
+    xenstore_write("1", XS_LID_EVENT_PATH);
+    //notify_com_citrix_xenclient_xcpmd_lid_changed(xcdbus_conn, XCPMD_SERVICE, XCPMD_PATH);
```

Comment:
This call doesn't yet exist in the IDL. I can go ahead and add it to the [existing pull request](https://github.com/OpenXT/idl/pull/7), if you think we stand to benefit from it.